docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,365 @@
+
+
+Usage and Impact
+----------------
+
+This is an optional feature, but should help nearly every application of GT4Py, since almost all partial differential equations at some point have a non-periodic boundary condition or require special treatment somewhere in the domain.
+
+
+Backward Compatibility
+----------------------
+
+This GDP aims to be fully backward-compatible.
+
+
+Detailed Description
+--------------------
+
+The intent of this GDP is to a specify horizontal bounds inside the stencil in a similar fashion to how vertical intervals are specified.
+However, the fact that there are multiple dimensions in the horizontal, halos, and domain decomposition imply the feature needs to be more complex than a simple index range used for the vertical.
+
+As introduced above, we propose adding a new gtscript `region()` specifier that specializes the stencil on a region of the space when a non-zero intersection with the global region occurs.
+The idea behind using the region objects is that they can incorporate both the iteration space a process sees and the global partitioning that an application must handle, all while hiding the complexity at the stencil level for model developers.
+In this proposal, the application remains in control of the partitioning data so **the partitioner is entirely transparent to both GT4Py and its backend.**
+
+
+Specification Options
++++++++++++++++++++++
+
+An issue we faced when designing the iteration space featiure is that the stages (`with interval()`) are, by the execution model specification, free to be re-ordered by the runtime.
+We propose adding the `region()` to **either** (1) the multistage, in which case the execution order is specified by GridTool's execution model **or** (2) within a stage.
+We plan to implement these in order, first finishing an implementation of (1) before attempting the second.
+If we agree to try (2) the example would look like:
+
+.. code-block:: python
+
+    @gtscript.stencil(backend=backend_type)
+    def stencil(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float, dt4: float):
+        with computation(PARALLEL), interval(...):
+            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
+            with region(WestEdge(0, 1), EastEdge(0, 1)):
+                ub = dt5 * (ut[0, -1, 0] + ut)
+            with region(SouthEdge(0, 1), NorthEdge(0, 1)):
+                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+This creates a switch-like stage where the execution at the point choses one of these options.
+The first line(s), or the lines ouside a `with` block, specify the default operation, and specializations are inside the blocks.
+An ambiguity arises when the same point is represented in multiple specializations, but that is easily cleared up by specifying that the **last** specialization is the one that takes precedence.
```

Comment:
I find the "take precedence" a little confusing, since it may mean "executed first". Would this sentece be equivalent to "specializations appear as executed in program order"?

Reply:
Yes, that's a better way of putting it. I'll update the text!